File: rotateImage.py

# problem: https://leetcode.com/problems/rotate-image/
# Runtime: 36 ms, faster than 61.41% of Python3 online submissions for Rotate Image.
# Memory Usage: 14.3 MB, less than 99.99% of Python3 online submissions for Rotate Image.
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """

        size_matrix = len(matrix)
        half_size_matrix = (size_matrix-1)/2
        max_i = size_matrix - 1
        for i in range(size_matrix):
            for j in range(i):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
        for i in range(size_matrix):
            for j in range(size_matrix/2):
                matrix[i][j], matrix[i][max_i -
                                        j] = matrix[i][max_i-j], matrix[i][j]

File: rotateImage.go
// problem: https://leetcode.com/problems/rotate-image/
// Runtime: 0 ms, faster than 100.00% of Go online submissions for Rotate Image.
// Memory Usage: 2.2 MB, less than 16.76% of Go online submissions for Rotate Image.

package algorithms

func rotate(matrix [][]int) {
	n := len(matrix)
	for i := 0; i < n; i++ {
		for j := 0; j < i; j++ {
			matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
		}
	}

	for i := 0; i < n; i++ {
		for j := 0; j < n/2; j++ {
			matrix[i][j], matrix[i][n-j-1] = matrix[i][n-j-1], matrix[i][j]
		}
	}
}
